test/config_cat/hooks_test.exs

```diff
@@ -0,0 +1,216 @@
+
+  @moduletag capture_log: true
+
+  @config ~J"""
+  {
+    "p": {
+      "u": "https://cdn-global.configcat.com",
+      "r": 0
+    },
+    "f": {
+      "testBoolKey": {"v": true,"t": 0, "p": [],"r": []},
+      "testStringKey": {"v": "testValue", "i": "id", "t": 1, "p": [],"r": [
+        {"i":"id1","v":"fake1","a":"Identifier","t":2,"c":"@test1.com"},
+        {"i":"id2","v":"fake2","a":"Identifier","t":2,"c":"@test2.com"}
+      ]},
+      "testIntKey": {"v": 1,"t": 2, "p": [],"r": []},
+      "testDoubleKey": {"v": 1.1,"t": 3,"p": [],"r": []},
+      "key1": {"v": true, "i": "fakeId1","p": [], "r": []},
+      "key2": {"v": false, "i": "fakeId2","p": [], "r": []}
+    }
+  }
+  """
+  @policy CachePolicy.manual()
+
+  defmodule TestHooks do
+    @moduledoc false
+    alias ConfigCat.Config
+    alias ConfigCat.EvaluationDetails
+
+    @spec on_client_ready(pid()) :: :ok
+    def on_client_ready(test_pid) do
+      send(test_pid, :on_client_ready)
+      :ok
+    end
+
+    @spec on_config_changed(Config.t(), pid()) :: :ok
+    def on_config_changed(config, test_pid) do
+      send(test_pid, {:on_config_changed, config})
+      :ok
+    end
+
+    @spec on_flag_evaluated(EvaluationDetails.t(), pid()) :: :ok
+    def on_flag_evaluated(%EvaluationDetails{} = details, test_pid) do
+      send(test_pid, {:on_flag_evaluated, details})
+      :ok
+    end
+
+    @spec on_error(String.t(), pid()) :: :ok
+    def on_error(message, test_pid) do
+      send(test_pid, {:on_error, message})
+      :ok
+    end
+  end
+
+  test "calls initial hooks" do
+    test_pid = self()
+
+    {:ok, instance_id} =
+      start_hooks(
+        on_client_ready: {TestHooks, :on_client_ready, [test_pid]},
+        on_config_changed: {TestHooks, :on_config_changed, [test_pid]},
+        on_flag_evaluated: {TestHooks, :on_flag_evaluated, [test_pid]},
+        on_error: {TestHooks, :on_error, [test_pid]}
+      )
+
+    :ok = start_client(instance_id: instance_id)
+
+    value = ConfigCat.get_value("testStringKey", "", client: instance_id)
+
+    {:ok, settings} = Config.fetch_settings(@config)
+
+    assert value == "testValue"
+    assert_received :on_client_ready
+    assert_received {:on_config_changed, ^settings}
+    assert_received {:on_flag_evaluated, _details}
+
+    refute_received {:on_error, _error}
+    refute_received _any_other_messages
+  end
+
+  test "calls subscribed hooks" do
+    test_pid = self()
+
+    {:ok, instance_id} = start_hooks()
+
+    _hooks =
+      ConfigCat.hooks(client: instance_id)
+      |> Hooks.add_on_client_ready({TestHooks, :on_client_ready, [test_pid]})
+      |> Hooks.add_on_config_changed({TestHooks, :on_config_changed, [test_pid]})
+      |> Hooks.add_on_flag_evaluated({TestHooks, :on_flag_evaluated, [test_pid]})
+      |> Hooks.add_on_error({TestHooks, :on_error, [test_pid]})
+
+    :ok = start_client(instance_id: instance_id)
+
+    value = ConfigCat.get_value("testStringKey", "", client: instance_id)
```

Comment:
This line occasionally fails in local development. There seems to be some kind of race condition that happens that I haven't yet figured out. I'll keep an eye on it and see what I can learn.